### DISCUSSION TITLE
Replace html element to page number

### DISCUSSION THREAD

**Discussion by acgtwentyone** [dompdf/dompdf#2951] [General]
Is it possible to use dompdf api to target a html element and then replace it by page number ? Let's say by using something like callbacks.

Ex: using something like this

```php
$dompdf->setCallbacks(
    [
        [
            'event' => 'begin_frame',
            'f' => function ($frame) {
                $node = $frame->get_node();
                if ($node->nodeName === 'h1') {
                    $GLOBALS['SECTION'] = $node->textContent;
                }
            }
        ],
        [
            'event' => 'end_page_render',
            'f' => function ($frame, $canvas) use ($fontMetrics) {
                $text = $GLOBALS['SECTION'];
                $font = $fontMetrics->getFont('monospace');
                $pageWidth = $canvas->get_width();
                $pageHeight = $canvas->get_height();
                $size = 12;
                $width = $fontMetrics->getTextWidth($text, $font, $size);
                $canvas->text($pageWidth - $width - 20, $pageHeight - 20, $text, $font, $size);
            }
        ]
    ]
);
$dompdf->render();
$dompdf->stream();
```

![Screenshot 2022-07-11 at 15 54 01](https://user-images.githubusercontent.com/94224806/178319013-4eabc40f-3720-4e11-bb6f-9ca387f1255e.png)

As you can see in the picture, let's say i want to replace "#" by 1/5. So we get "Página 1/5".

**Comment by fredsal** [comment]
Did you read other discussions and issues about this?

**Reply by acgtwentyone** [reply]
Yes. I couldn't find anything like that.

**Reply by fredsal** [reply]
anything, seriusly??
https://github.com/dompdf/dompdf/issues/1636
https://github.com/dompdf/dompdf/issues/1254
https://github.com/dompdf/dompdf/discussions/2860
https://github.com/barryvdh/laravel-dompdf/issues/37
https://github.com/barryvdh/laravel-dompdf/issues/452
https://github.com/barryvdh/laravel-dompdf/issues/352
[generated-content](https://github.com/dompdf/dompdf/wiki/Common-Use-Cases#generated-content)

**Reply by acgtwentyone** [reply]
> anything, seriusly?? #1636 #1254 #2860 [barryvdh/laravel-dompdf#37](https://github.com/barryvdh/laravel-dompdf/issues/37) [barryvdh/laravel-dompdf#452](https://github.com/barryvdh/laravel-dompdf/issues/452) [barryvdh/laravel-dompdf#352](https://github.com/barryvdh/laravel-dompdf/issues/352) [generated-content](https://github.com/dompdf/dompdf/wiki/Common-Use-Cases#generated-content)

Thank you @fredsal. I's solved by this #1636 as you pointed me out.

**Reply by parallels999** [reply]
@fredsal there is people too lazy to search

**Reply by acgtwentyone** [reply]
> @fredsal there is people too lazy to search

Thank you @parallels999. If you can't help, don't say things you don't know.

**Reply by fredsal** [reply]
@parallels999 it happens a lot 😄

**Reply by acgtwentyone** [reply]
Thanks for each one of you. I should have search more.

**Comment by bsweeney** [comment]
You _could_ do something like that, though if you're just inserting the standard page number for each page you can use [generated content](https://github.com/dompdf/dompdf/wiki/Common-Use-Cases#generated-content).

**Reply by acgtwentyone** [reply]
Thank you for pointing me out. I'll take a look.

**Reply by acgtwentyone** [reply]
> You _could_ do something like that, though if you're just inserting the standard page number for each page you can use [generated content](https://github.com/dompdf/dompdf/wiki/Common-Use-Cases#generated-content).

@bsweeney I actually want to replace the text "#" to the current page number. Lets say 1/5 for example. So instead of "Pagina #" we get "Pagina 1/5". Is it possible to do using dompdf ?

**Reply by bsweeney** [reply]
Per the link, you can get the current page using just CSS. Or using script but why go through the trouble.

Unfortunately, the total number of pages is not available until the document is fully rendered. This is due to the way that Dompdf renders documents. Typically in that scenario we recommend writing the text as part of a [page script](https://github.com/dompdf/dompdf/wiki/Usage#canvas).

It is possible to do string replacement after the document is fully rendered, though it's primarily a hack around current functionality. At any rate, I see @fredsal (thanks, buddy!) gave you some links that you were able to use to solve your issue.

**Reply by acgtwentyone** [reply]
> Per the link, you can get the current page using just CSS. Or using script but why go through the trouble.
> 
> Unfortunately, the total number of pages is not available until the document is fully rendered. This is due to the way that Dompdf renders documents. Typically in that scenario we recommend writing the text as part of a [page script](https://github.com/dompdf/dompdf/wiki/Usage#canvas).
> 
> It is possible to do string replacement after the document is fully rendered, though it's primarily a hack around current functionality. At any rate, I see @fredsal (thanks, buddy!) gave you some links that you were able to use to solve your issue.

Thank you.
